File: generate_icon.py
#!/usr/bin/env python3
"""Generate a retro Windows 3.1-style Notepad icon at multiple sizes."""

from PIL import Image, ImageDraw
import os

def draw_notepad_icon(size):
    """Draw a retro notepad icon reminiscent of the classic Windows Notepad."""
    img = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)

    s = size / 512.0  # scale factor

    # Slight shadow behind the notepad
    shadow_offset = int(6 * s)
    shadow_coords = [
        int(80*s) + shadow_offset, int(30*s) + shadow_offset,
        int(430*s) + shadow_offset, int(480*s) + shadow_offset
    ]
    draw.rectangle(shadow_coords, fill=(0, 0, 0, 60))

    # Main notepad body - cream/yellow paper
    paper_color = (255, 255, 220)
    paper_left = int(80*s)
    paper_top = int(30*s)
    paper_right = int(430*s)
    paper_bottom = int(480*s)
    draw.rectangle([paper_left, paper_top, paper_right, paper_bottom], fill=paper_color)

    # 3D raised border (Win3.1 style)
    border_w = max(int(3*s), 1)
    # Top/left highlight (white)
    for i in range(border_w):
        draw.line([(paper_left-i, paper_bottom+i), (paper_left-i, paper_top-i)], fill=(255, 255, 255), width=1)
        draw.line([(paper_left-i, paper_top-i), (paper_right+i, paper_top-i)], fill=(255, 255, 255), width=1)
    # Bottom/right shadow (dark gray)
    for i in range(border_w):
        draw.line([(paper_right+i, paper_top-i), (paper_right+i, paper_bottom+i)], fill=(80, 80, 80), width=1)
        draw.line([(paper_right+i, paper_bottom+i), (paper_left-i, paper_bottom+i)], fill=(80, 80, 80), width=1)

    # Blue title bar at top of notepad
    title_top = paper_top
    title_bottom = paper_top + int(45*s)
    title_color = (0, 0, 128)  # Classic dark blue
    draw.rectangle([paper_left, title_top, paper_right, title_bottom], fill=title_color)

    # Title bar text - draw simple pixel "Notepad" text
    text_y = title_top + int(10*s)
    text_x = paper_left + int(15*s)
    text_h = int(22*s)
    text_color = (255, 255, 255)

    # Simple block letters "NOTE" (retro pixel style)
    char_w = int(18*s)
    char_gap = int(4*s)

    def draw_block_char(x, y, ch, h, w):
        """Draw simplified block characters."""
        t = max(int(3*s), 1)  # stroke thickness
        if ch == 'N':
            draw.rectangle([x, y, x+t, y+h], fill=text_color)
            draw.rectangle([x+w-t, y, x+w, y+h], fill=text_color)
            # diagonal
            for i in range(h):
                dx = int(i * w / h)
                draw.rectangle([x+dx, y+i, x+dx+t, y+i+t], fill=text_color)
        elif ch == 'O':
            draw.rectangle([x, y, x+w, y+t], fill=text_color)
            draw.rectangle([x, y+h-t, x+w, y+h], fill=text_color)
            draw.rectangle([x, y, x+t, y+h], fill=text_color)
            draw.rectangle([x+w-t, y, x+w, y+h], fill=text_color)
        elif ch == 'T':
            draw.rectangle([x, y, x+w, y+t], fill=text_color)
            draw.rectangle([x+w//2-t//2, y, x+w//2+t//2+t, y+h], fill=text_color)
        elif ch == 'E':
            draw.rectangle([x, y, x+w, y+t], fill=text_color)
            draw.rectangle([x, y+h//2-t//2, x+w-int(3*s), y+h//2+t//2], fill=text_color)
            draw.rectangle([x, y+h-t, x+w, y+h], fill=text_color)
            draw.rectangle([x, y, x+t, y+h], fill=text_color)
        elif ch == 'P':
            draw.rectangle([x, y, x+t, y+h], fill=text_color)
            draw.rectangle([x, y, x+w, y+t], fill=text_color)
            draw.rectangle([x, y+h//2-t//2, x+w, y+h//2+t//2], fill=text_color)
            draw.rectangle([x+w-t, y, x+w, y+h//2], fill=text_color)
        elif ch == 'A':
            draw.rectangle([x, y, x+w, y+t], fill=text_color)
            draw.rectangle([x, y, x+t, y+h], fill=text_color)
            draw.rectangle([x+w-t, y, x+w, y+h], fill=text_color)
            draw.rectangle([x, y+h//2-t//2, x+w, y+h//2+t//2], fill=text_color)
        elif ch == 'D':
            draw.rectangle([x, y, x+t, y+h], fill=text_color)
            draw.rectangle([x, y, x+w-int(5*s), y+t], fill=text_color)
            draw.rectangle([x, y+h-t, x+w-int(5*s), y+h], fill=text_color)
            # curved right side approximation
            draw.rectangle([x+w-t, y+int(5*s), x+w, y+h-int(5*s)], fill=text_color)
            draw.rectangle([x+w-int(5*s)-t, y, x+w-int(5*s), y+int(5*s)], fill=text_color)
            draw.rectangle([x+w-int(5*s)-t, y+h-int(5*s), x+w-int(5*s), y+h], fill=text_color)

    for i, ch in enumerate("NOTEPAD"):
        draw_block_char(text_x + i * (char_w + char_gap), text_y, ch, text_h, char_w)

    # Lined paper - horizontal blue lines
    line_color = (180, 200, 230)
    content_top = title_bottom + int(15*s)
    line_spacing = int(28*s)
    margin_left = paper_left + int(20*s)
    margin_right = paper_right - int(20*s)

    line_y = content_top
    while line_y < paper_bottom - int(25*s):
        draw.line([(margin_left, line_y), (margin_right, line_y)], fill=line_color, width=max(int(1.5*s), 1))
        line_y += line_spacing

    # Red margin line (classic ruled paper)
    red_margin_x = paper_left + int(55*s)
    draw.line([(red_margin_x, content_top - int(10*s)), (red_margin_x, paper_bottom - int(10*s))],
              fill=(220, 80, 80), width=max(int(2*s), 1))

    # Simulated text scribbles on the lines
    scribble_color = (40, 40, 40)
    scribble_h = max(int(3*s), 1)
    text_left = red_margin_x + int(15*s)

    # Different length "text" blocks on each line
    line_y = content_top
    text_patterns = [0.85, 0.6, 0.75, 0.4, 0.9, 0.55, 0.7, 0.3, 0.65, 0.8, 0.5, 0.45]
    pattern_idx = 0
    while line_y < paper_bottom - int(25*s):
        width_frac = text_patterns[pattern_idx % len(text_patterns)]
        text_width = int((margin_right - text_left) * width_frac)

        # Draw text as small dashes to look like handwriting
        tx = text_left
        while tx < text_left + text_width:
            word_len = int((8 + (pattern_idx * 7 + tx) % 20) * s)
            word_end = min(tx + word_len, text_left + text_width)
            draw.rectangle([tx, line_y - int(8*s), word_end, line_y - int(8*s) + scribble_h], fill=scribble_color)
            tx = word_end + int(6*s)

        line_y += line_spacing
        pattern_idx += 1

    # Pencil laying diagonally across bottom-right
    pencil_color = (220, 180, 50)       # Yellow pencil body
    pencil_dark = (180, 140, 30)        # Darker edge
    pencil_tip = (60, 50, 40)           # Dark tip
    pencil_eraser = (230, 120, 130)     # Pink eraser
    pencil_band = (180, 180, 180)       # Metal band

    # Pencil coordinates (diagonal from bottom-left area to right)
    px1, py1 = int(200*s), int(500*s)   # eraser end
    px2, py2 = int(470*s), int(360*s)   # tip end

    pencil_w = int(18*s)

    # Calculate pencil direction
    import math
    dx = px2 - px1
    dy = py2 - py1
    length = math.sqrt(dx*dx + dy*dy)
    nx = -dy / length * pencil_w / 2  # normal x
    ny = dx / length * pencil_w / 2   # normal y

    # Pencil body polygon
    body_start = 0.12  # where body starts (after eraser)
    body_end = 0.88    # where body ends (before tip)

    bx1 = px1 + dx * body_start
    by1 = py1 + dy * body_start
    bx2 = px1 + dx * body_end
    by2 = py1 + dy * body_end

    body_poly = [
        (bx1 + nx, by1 + ny),
        (bx2 + nx, by2 + ny),
        (bx2 - nx, by2 - ny),
        (bx1 - nx, by1 - ny),
    ]
    draw.polygon(body_poly, fill=pencil_color, outline=pencil_dark)

    # Eraser
    ex1, ey1 = px1, py1
    ex2, ey2 = bx1, by1
    eraser_poly = [
        (ex1 + nx*0.8, ey1 + ny*0.8),
        (ex2 + nx, ey2 + ny),
        (ex2 - nx, ey2 - ny),
        (ex1 - nx*0.8, ey1 - ny*0.8),
    ]
    draw.polygon(eraser_poly, fill=pencil_eraser)

    # Metal band between eraser and body
    band_start = body_start - 0.02
    band_end = body_start + 0.02
    mbx1 = px1 + dx * band_start
    mby1 = py1 + dy * band_start
    mbx2 = px1 + dx * band_end
    mby2 = py1 + dy * band_end
    band_poly = [
        (mbx1 + nx, mby1 + ny),
        (mbx2 + nx, mby2 + ny),
        (mbx2 - nx, mby2 - ny),
        (mbx1 - nx, mby1 - ny),
    ]
    draw.polygon(band_poly, fill=pencil_band)

    # Pencil tip (triangle)
    tip_poly = [
        (bx2 + nx, by2 + ny),
        (px2, py2),
        (bx2 - nx, by2 - ny),
    ]
    draw.polygon(tip_poly, fill=(240, 220, 180))

    # Dark point at very tip
    tip_r = int(4*s)
    draw.ellipse([px2-tip_r, py2-tip_r, px2+tip_r, py2+tip_r], fill=pencil_tip)

    return img


def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    icon_dir = os.path.join(script_dir, "MacNotepad", "Resources", "Assets.xcassets", "AppIcon.appiconset")

    sizes = {
        "icon_16x16.png": 16,
        "icon_32x32.png": 32,
        "icon_64x64.png": 64,
        "icon_128x128.png": 128,
        "icon_256x256.png": 256,
        "icon_512x512.png": 512,
        "icon_1024x1024.png": 1024,
    }

    for filename, size in sizes.items():
        print(f"Generating {filename} ({size}x{size})...")
        img = draw_notepad_icon(size)
        img.save(os.path.join(icon_dir, filename))

    print("Done! Icon files generated.")


if __name__ == "__main__":
    main()

File: generate_icon.swift
#!/usr/bin/env swift

import Cocoa

func drawNotepadIcon(size: Int) -> NSImage {
    let s = CGFloat(size) / 512.0
    let img = NSImage(size: NSSize(width: size, height: size))

    img.lockFocus()

    let ctx = NSGraphicsContext.current!.cgContext
    // Flip to top-left origin
    ctx.translateBy(x: 0, y: CGFloat(size))
    ctx.scaleBy(x: 1, y: -1)

    // Clear
    ctx.clear(CGRect(x: 0, y: 0, width: size, height: size))

    // Shadow
    let so = 6 * s
    ctx.setFillColor(NSColor(white: 0, alpha: 0.15).cgColor)
    ctx.fill(CGRect(x: 80*s + so, y: 30*s + so, width: 350*s, height: 450*s))

    // Paper body
    let pl = 80*s, pt = 30*s, pw = 350*s, ph = 450*s
    ctx.setFillColor(NSColor(red: 1, green: 1, blue: 0.86, alpha: 1).cgColor)
    ctx.fill(CGRect(x: pl, y: pt, width: pw, height: ph))

    // 3D border
    let bw = max(3*s, 1)
    // White top/left
    ctx.setStrokeColor(NSColor.white.cgColor)
    ctx.setLineWidth(bw)
    ctx.move(to: CGPoint(x: pl - bw/2, y: pt + ph + bw/2))
    ctx.addLine(to: CGPoint(x: pl - bw/2, y: pt - bw/2))
    ctx.addLine(to: CGPoint(x: pl + pw + bw/2, y: pt - bw/2))
    ctx.strokePath()
    // Dark bottom/right
    ctx.setStrokeColor(NSColor(white: 0.31, alpha: 1).cgColor)
    ctx.move(to: CGPoint(x: pl + pw + bw/2, y: pt - bw/2))
    ctx.addLine(to: CGPoint(x: pl + pw + bw/2, y: pt + ph + bw/2))
    ctx.addLine(to: CGPoint(x: pl - bw/2, y: pt + ph + bw/2))
    ctx.strokePath()

    // Blue title bar
    let tt = pt, tb = pt + 45*s
    ctx.setFillColor(NSColor(red: 0, green: 0, blue: 0.5, alpha: 1).cgColor)
    ctx.fill(CGRect(x: pl, y: tt, width: pw, height: 45*s))

    // Title text "NOTEPAD" as simple block letters
    let tc = NSColor.white.cgColor
    let charH = 20*s
    let stroke = max(3*s, 1)
    let charW = 18*s
    let gap = 4*s
    let textY = tt + 12*s
    let textX = pl + 15*s

    func rect(_ x: CGFloat, _ y: CGFloat, _ w: CGFloat, _ h: CGFloat) {
        ctx.setFillColor(tc)
        ctx.fill(CGRect(x: x, y: y, width: w, height: h))
    }

    func drawChar(_ ch: Character, _ x: CGFloat, _ y: CGFloat) {
        let w = charW, h = charH, t = stroke
        switch ch {
        case "N":
            rect(x, y, t, h); rect(x+w-t, y, t, h)
            for i in stride(from: CGFloat(0), to: h, by: 1) {
                let dx = i * w / h
                rect(x+dx, y+i, t, 1)
            }
        case "O":
            rect(x, y, w, t); rect(x, y+h-t, w, t)
            rect(x, y, t, h); rect(x+w-t, y, t, h)
        case "T":
            rect(x, y, w, t); rect(x+w/2-t/2, y, t, h)
        case "E":
            rect(x, y, w, t); rect(x, y+h/2, w-3*s, t)
            rect(x, y+h-t, w, t); rect(x, y, t, h)
        case "P":
            rect(x, y, t, h); rect(x, y, w, t)
            rect(x, y+h/2, w, t); rect(x+w-t, y, t, h/2+t)
        case "A":
            rect(x, y, w, t); rect(x, y, t, h)
            rect(x+w-t, y, t, h); rect(x, y+h/2, w, t)
        case "D":
            rect(x, y, t, h); rect(x, y, w-5*s, t)
            rect(x, y+h-t, w-5*s, t); rect(x+w-t, y+5*s, t, h-10*s)
        default: break
        }
    }

    for (i, ch) in "NOTEPAD".enumerated() {
        drawChar(ch, textX + CGFloat(i) * (charW + gap), textY)
    }

    // Lined paper
    let contentTop = tb + 15*s
    let lineSpacing = 28*s
    let ml = pl + 20*s, mr = pl + pw - 20*s

    ctx.setStrokeColor(NSColor(red: 0.71, green: 0.78, blue: 0.9, alpha: 1).cgColor)
    ctx.setLineWidth(max(1.5*s, 1))
    var ly = contentTop
    while ly < pt + ph - 25*s {
        ctx.move(to: CGPoint(x: ml, y: ly))
        ctx.addLine(to: CGPoint(x: mr, y: ly))
        ctx.strokePath()
        ly += lineSpacing
    }

    // Red margin line
    let rmx = pl + 55*s
    ctx.setStrokeColor(NSColor(red: 0.86, green: 0.31, blue: 0.31, alpha: 1).cgColor)
    ctx.setLineWidth(max(2*s, 1))
    ctx.move(to: CGPoint(x: rmx, y: contentTop - 10*s))
    ctx.addLine(to: CGPoint(x: rmx, y: pt + ph - 10*s))
    ctx.strokePath()

    // Text scribbles
    let textLeft = rmx + 15*s
    let scribbleH = max(3*s, 1)
    let pats: [CGFloat] = [0.85, 0.6, 0.75, 0.4, 0.9, 0.55, 0.7, 0.3, 0.65, 0.8, 0.5, 0.45]
    ctx.setFillColor(NSColor(white: 0.16, alpha: 1).cgColor)
    ly = contentTop
    var pi = 0
    while ly < pt + ph - 25*s {
        let tw = (mr - textLeft) * pats[pi % pats.count]
        var tx = textLeft
        while tx < textLeft + tw {
            let wl = (8 + CGFloat((pi * 7 + Int(tx)) % 20)) * s
            let we = min(tx + wl, textLeft + tw)
            ctx.fill(CGRect(x: tx, y: ly - 8*s, width: we - tx, height: scribbleH))
            tx = we + 6*s
        }
        ly += lineSpacing
        pi += 1
    }

    // Pencil (diagonal)
    let px1 = 200*s, py1 = 500*s
    let px2 = 470*s, py2 = 360*s
    let pencilW = 18*s
    let dx = px2 - px1, dy = py2 - py1
    let length = sqrt(dx*dx + dy*dy)
    let nx = -dy / length * pencilW / 2
    let ny = dx / length * pencilW / 2

    let bs: CGFloat = 0.12, be: CGFloat = 0.88
    let bx1 = px1 + dx*bs, by1 = py1 + dy*bs
    let bx2 = px1 + dx*be, by2 = py1 + dy*be

    // Body
    ctx.setFillColor(NSColor(red: 0.86, green: 0.71, blue: 0.2, alpha: 1).cgColor)
    ctx.move(to: CGPoint(x: bx1+nx, y: by1+ny))
    ctx.addLine(to: CGPoint(x: bx2+nx, y: by2+ny))
    ctx.addLine(to: CGPoint(x: bx2-nx, y: by2-ny))
    ctx.addLine(to: CGPoint(x: bx1-nx, y: by1-ny))
    ctx.closePath()
    ctx.fillPath()

    // Darker outline
    ctx.setStrokeColor(NSColor(red: 0.71, green: 0.55, blue: 0.12, alpha: 1).cgColor)
    ctx.setLineWidth(max(s, 0.5))
    ctx.move(to: CGPoint(x: bx1+nx, y: by1+ny))
    ctx.addLine(to: CGPoint(x: bx2+nx, y: by2+ny))
    ctx.addLine(to: CGPoint(x: bx2-nx, y: by2-ny))
    ctx.addLine(to: CGPoint(x: bx1-nx, y: by1-ny))
    ctx.closePath()
    ctx.strokePath()

    // Eraser
    ctx.setFillColor(NSColor(red: 0.9, green: 0.47, blue: 0.51, alpha: 1).cgColor)
    ctx.move(to: CGPoint(x: px1+nx*0.8, y: py1+ny*0.8))
    ctx.addLine(to: CGPoint(x: bx1+nx, y: by1+ny))
    ctx.addLine(to: CGPoint(x: bx1-nx, y: by1-ny))
    ctx.addLine(to: CGPoint(x: px1-nx*0.8, y: py1-ny*0.8))
    ctx.closePath()
    ctx.fillPath()

    // Metal band
    let mbs: CGFloat = 0.10, mbe: CGFloat = 0.14
    let mbx1 = px1+dx*mbs, mby1 = py1+dy*mbs
    let mbx2 = px1+dx*mbe, mby2 = py1+dy*mbe
    ctx.setFillColor(NSColor(white: 0.7, alpha: 1).cgColor)
    ctx.move(to: CGPoint(x: mbx1+nx, y: mby1+ny))
    ctx.addLine(to: CGPoint(x: mbx2+nx, y: mby2+ny))
    ctx.addLine(to: CGPoint(x: mbx2-nx, y: mby2-ny))
    ctx.addLine(to: CGPoint(x: mbx1-nx, y: mby1-ny))
    ctx.closePath()
    ctx.fillPath()

    // Tip
    ctx.setFillColor(NSColor(red: 0.94, green: 0.86, blue: 0.71, alpha: 1).cgColor)
    ctx.move(to: CGPoint(x: bx2+nx, y: by2+ny))
    ctx.addLine(to: CGPoint(x: px2, y: py2))
    ctx.addLine(to: CGPoint(x: bx2-nx, y: by2-ny))
    ctx.closePath()
    ctx.fillPath()

    // Dark point
    let tr = 4*s
    ctx.setFillColor(NSColor(red: 0.24, green: 0.2, blue: 0.16, alpha: 1).cgColor)
    ctx.fillEllipse(in: CGRect(x: px2-tr, y: py2-tr, width: tr*2, height: tr*2))

    img.unlockFocus()
    return img
}

func savePNG(_ image: NSImage, to path: String, size: Int) {
    let rep = NSBitmapImageRep(bitmapDataPlanes: nil,
                                pixelsWide: size, pixelsHigh: size,
                                bitsPerSample: 8, samplesPerPixel: 4,
                                hasAlpha: true, isPlanar: false,
                                colorSpaceName: .deviceRGB,
                                bytesPerRow: 0, bitsPerPixel: 0)!
    NSGraphicsContext.saveGraphicsState()
    NSGraphicsContext.current = NSGraphicsContext(bitmapImageRep: rep)
    image.draw(in: NSRect(x: 0, y: 0, width: size, height: size))
    NSGraphicsContext.restoreGraphicsState()

    let data = rep.representation(using: .png, properties: [:])!
    try! data.write(to: URL(fileURLWithPath: path))
}

let iconDir = "MacNotepad/Resources/Assets.xcassets/AppIcon.appiconset"
let sizes = [
    ("icon_16x16.png", 16),
    ("icon_32x32.png", 32),
    ("icon_64x64.png", 64),
    ("icon_128x128.png", 128),
    ("icon_256x256.png", 256),
    ("icon_512x512.png", 512),
    ("icon_1024x1024.png", 1024),
]

for (name, size) in sizes {
    print("Generating \(name)...")
    let img = drawNotepadIcon(size: size)
    savePNG(img, to: "\(iconDir)/\(name)", size: size)
}
print("Done!")
